flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java

```diff
@@ -100,14 +92,24 @@
 	/** Flag indicating whether this group has been closed. */
 	private volatile boolean closed;
 
+	private final InternalMetricScope scope;
+
 	// ------------------------------------------------------------------------
 
 	public AbstractMetricGroup(MetricRegistry registry, String[] scope, A parent) {
 		this.registry = checkNotNull(registry);
-		this.scopeComponents = checkNotNull(scope);
 		this.parent = parent;
-		this.scopeStrings = new String[registry.getNumberReporters()];
 		this.logicalScopeStrings = new String[registry.getNumberReporters()];
+		this.scope = new InternalMetricScope(
+			registry,
+			scope,
+			this::getAllVariables
+		);
+	}
+
+	@Override
+	public InternalMetricScope getScope() {
+		return scope;
 	}
 
 	public Map<String, String> getAllVariables() {
```

Comment:
This method is in fact `@Override`. My suggestion would be to rename it (and make `private`), so that there is only one "code path" to get those variables (via `scope.getAllVariables()`).
A new name could be `computeAllVariables()`.

Also, please see my top level comment.